Enable caching + add new non-admin endpoints to UX backend server

SanjalKatiyar opened the pull request:
https://issues.redhat.com/browse/RHSTOR-7965
https://issues.redhat.com/browse/RHSTOR-7970
https://issues.redhat.com/browse/RHSTOR-8003

### Introduced Changes:

- Segregated 2 paths for OAuth Proxy:
   - **"/":** (Default) Only authenticated + authorized users allowed. This path is needed for admin user APIs.
   - **"/info/":** Only authenticated users allowed (no authorization). This path is needed for any OpenShift user (admin or non-admin) APIs.

- Enabled cached client:
   - `DefaultNamespaces` set to watch `openshift-storage` and `openshift-storage-extended` only.
   - Indexing (cluster-scoped resources).
 
- Added 3 new endpoints to "/info/" path:
   - **"/info/featureflags?flags=noobaa,rgw":**
     - Enable/Disable (or Show/Hide) features on UI based on flags.
     - Will be used globally by UI (irrespective of user type or which page user navigate to).
     - Faster & lightweight (less response size, we don't need all the info, we only need to know the subset).
   - **"/info/bucket/\<bucket-name\>":**
     - Determine whether a S3 bucket (NooBaa/RGW) is created via OBC or directly via S3 API.
     - Specific non-admin only use case for now.
   - **"/info/storages":**
     - Comprehensive namespace-wise cluster state.
     - Specific non-admin only use case for now. Also, only called when/if user go to a specific UI page (local usage), so only needed sometimes optionally.
     - **Comparatively** slower and comprehensive (we need all the information for UI to make any decision, so response is **relatively** bigger).

### Testing (endpoint):
```
$ curl -k -H "Authorization: Bearer <ADMIN_USER>" https://ux-backend-proxy.openshift
-storage.svc.cluster.local:8888/info/featureflags?flags=noobaa,rgw,fakeFlag

>> {"fakeFlag":{"value":false,"error":"unsupported flag"},"noobaa":{"value":true},"rgw":{"value":false}}
```

```
$ curl -k -H "Authorization: Bearer <NON_ADMIN_USER>" https://ux-backend-proxy.openshift
-storage.svc.cluster.local:8888/info/featureflags?flags=noobaa,rgw,fakeFlag

>> {"fakeFlag":{"value":false,"error":"unsupported flag"},"noobaa":{"value":true},"rgw":{"value":false}}
```

```
$ curl -k -H "Authorization: Bearer <ADMIN_USER>" https://ux-backend-proxy.openshift
-storage.svc.cluster.local:8888/info/bucket/test-bucket

>> {"createdVia":"obc"}
```

```
$ curl -k -H "Authorization: Bearer <ADMIN_USER>" https://ux-backend-proxy.openshift
-storage.svc.cluster.local:8888/info/bucket/test-e60c8369-b6d9-4e84-bf2d-89cae86cbafd

>> {"createdVia":"s3"}
```

```
$ curl -k -H "Authorization: Bearer <NON_ADMIN_USER>" https://ux-backend-proxy.openshift
-storage.svc.cluster.local:8888/info/bucket/test-e60c8369-b6d9-4e84-bf2d-89cae86cbafd

>> {"createdVia":"s3"}
```

```
$ curl -v -k -H "Authorization: Bearer <FAKE_USER>" https://ux-backend-proxy.o
penshift-storage.svc.cluster.local:8888/info/featureflags?flags=noobaa,rgw,fakeFlag

>> HTTP/1.1 403 Forbidden
```

```
$ curl -v -k -H "Authorization: Bearer <FAKE_USER>" https://ux-backend-proxy.o
penshift-storage.svc.cluster.local:8888/info/bucket/test-bucket

>> HTTP/1.1 403 Forbidden
```

```
$ curl -v -k -H "Authorization: Bearer <NON_ADMIN_USER>" https://ux-backend-proxy.o
penshift-storage.svc.cluster.local:8888/info/storages

>> {"operatorNamespace":"openshift-storage","clusterNamespaces":{"openshift-storage":{"storageClusterName":"test","deploymentType":"internal"}}}
```

```
$ curl -v -k -H "Authorization: Bearer <NON_ADMIN_USER>" https://ux-backend-proxy.o
penshift-storage.svc.cluster.local:8888/info/storages

>> {"operatorNamespace":"openshift-storage","clusterNamespaces":{"openshift-storage":{"storageClusterName":"test","deploymentType":"internal"}}}
```

### Testing (cache):

First request - slower (populates cache)
```
{"operatorNamespace":"openshift-storage","clusterNamespaces":{"openshift-storage":{"storageClusterName":"test","deploymentType":"internal"}}}

real    0m0.459s
user    0m0.009s
sys     0m0.006s
```

Subsequent requests - faster (cache exists)
```
{"operatorNamespace":"openshift-storage","clusterNamespaces":{"openshift-storage":{"storageClusterName":"test-new","deploymentType":"internal"}}}

real    0m0.023s
user    0m0.010s
sys     0m0.004s
```

leelavg commented:
@SanjalKatiyar I didn't look at code yet, looking at PR description alone I'm not sure that should be the final, is there a finalized design doc for the API? Some of my questions are:

1. grouping 
- `/` vs `/auth` -> a client doesn't need to know this distinction unless the grouping is in terms of resources but not in terms of privilege levels
- I believe you chose this for ease of representation at oauth-proxy, does that not accept a mounted configmap?
- your api should define the endpoints and what operations possible on that and the middleware (proxy) need to take a decision whether a client should be allowed or not.

2. featureflags
- why should the pathparam (`/featureflags`) and query (`flags`) has redundant information?
- I would only use queries on a list for filtering the resources and not support the splitting of values etc, a simple flat structure could be good
- why can't we follow k8s design, like
`GET /flags` list all the currently available flags with their state
`PUT /flags/noobaa {"enable": true}` activate the flag or some more operations to be done
`PUT /flags/noobaa {"enable": false}` deactivate the flag
`PUT /flags/SOMETHING {"enable": true}` we don't support this flag
- a single user updating flag reflects the UI to all other logged in users, is that the intention or the UI should only be different to that user (I suppose the former)

3. bucket-origin
- better not make the API answer a question, this isn't something like a RPC function call 😅
- return the metadata for a bucket as the response against the name of the bucket? like
`GET /buckets` list all the buckets we are aware of, maybe you can limit to only the names?
`GET /buckets/{name}` respond with the metadata that you have for this bucket and let the client do the processing for any fields if it has to

Please excuse, I don't have all background information of the requirement and just provided general guidelines, open to be corrected!

SanjalKatiyar commented:
> however, grouping also has merits: 1. proxy normalizes the path, like /auth/../../ would be normalized to / and that review is checked, so it's kinda safe, 2. it is convenient to have all the prefixes checked in a single instance

Yes, instead of multiple redundant entries under `-openshift-delegate-urls` (one per endpoint), idea is to have just one entry and all endpoints which needs similar auth + authr can be defined using that prefix.

> I suggest to have a different name rather than auth, like tenant?

I am not sure, intention behind "/auth/" was that all endpoints under it will just authenticate (hence the prefix).
"/tenant/" could ideally have different access level, so that is not 100% accurate as well.
There could be other suggestions as well like using "/user/" or "/info/", but unless there is a strong objection against using "/auth/", IMHO let's stick to it.

> the other option is doing a POST w/ a body to check for all the required flags since the requirement is not to have multiple calls, unless we'll have more flags and the query value keeps on increasing hitting limits we are fine w/ GET /flags=a,b

We can/will definitely support more flags in future, but I don't think it will ever exceed URL limit (we don't need that many).
Since this endpoint doesn't need to support any "create" operations ("read" only retrievals is the intention here), let's stick to "GET" method only for now. We can always enhance if/when needed, UI is the only consumer of this endpoint.

> agreed to use params GET /bucket/{name} to return {"createdVia": "s3/obc"}
 
Updated @leelavg, PTAL.

leelavg commented on services/ux-backend/handlers/bucket/handler.go:
```diff
@@ -0,0 +1,97 @@
+
+const (
+	CreationMethodOBC = "obc"
+	CreationMethodS3  = "s3"
```
are these being exported?

SanjalKatiyar replied:
no...
should I use camelCase format for the variable name ??

SanjalKatiyar replied:
updated...

SanjalKatiyar commented:
> * to me it reads as something to do with authentication functionality like, some client trying to authenticate or get a token etc.

I could think of these and corresponding issues:
`/auth/` - As u mentioned, sounds like "something to do with authentication functionality".
`/tenant/` - AFAIK "tenant" in OpenShift is usually a namespace/project. Also, each tenant can theoretically have diff access levels (authorisation). So, this is misleading as well.
`/public/` or `/open/`- Since we are still authenticating this is not a public/open API.
`/user/` - Sound like user-specific API, in reality it's opposite.
`/info/` - Very generic term, any API can be used for information purpose. 
`/console/` - This is wrong as well. UI is the main/major consumer of entire ux-backend-server, not just single path under it.

That's why I just stuck with `/auth/`. But, let me know if any other suggestion ?? Or, if u still think `/tenant/` is the best prefix. I will update in that case @leelavg.

rewantsoni commented on services/ux-backend/handlers/featureflags/handler.go:
```diff
@@ -0,0 +1,215 @@
+	}
+}
+
+func checkNoobaaFlag(client ctrlclient.Client, namespace string, ctx context.Context) (bool, error) {
```
>If u r suggesting importing this utility into the "/featureflag" handler then: that won't work, since the StorageCluster instance (which this utility depends on) won't cover all possible/potential future use cases for this endpoint.

The method depends on storageCluster to generate the names for CephFilesystem and cephNFS based on storageCluster name and namespace. I think we might be able to reuse it since GetAvailableServices is a util designed to get the services available

If we don't want to reuse, then we know the name of the cephObjectStore and noobaa instance that we create, we don't need to list them we can go ahead and get it

leelavg commented on services/ux-backend/handlers/bucket/handler.go:
```diff
@@ -0,0 +1,94 @@
+	w.Header().Set("Allow", "GET")
+	w.WriteHeader(http.StatusMethodNotAllowed)
+
+	if _, err := w.Write([]byte(fmt.Sprintf("Unsupported method: %s", r.Method))); err != nil {
```
I believe there is a method in fmt package which outputs to io.Reader, like
fmt.Fprintf(w, "msg", ...args)?

SanjalKatiyar replied:
leaving it as it is for now, same as other (existing) handlers we have...

leelavg commented on services/ux-backend/handlers/bucket/handler.go:
```diff
@@ -0,0 +1,94 @@
+	}
+
+	w.Header().Set("Content-Type", "application/json")
+	w.WriteHeader(http.StatusOK)
```
I believe the first write automatically adds the OK header.

leelavg commented on services/ux-backend/handlers/bucket/handler.go:
```diff
@@ -0,0 +1,97 @@
+	klog.Infof("Checking if bucket '%s' was created via OBC", bucketName)
+	
+	objectBucketList := &nbv1alpha1.ObjectBucketList{}
+	if err := client.List(ctx, objectBucketList); err != nil {
```
> UI won't request data for all buckets at once.
- that's not my point, even if UI is requesting for a single bucket, backend is do a list of all buckets to find that single bucket. so w/o cache + index every request pulls total number of buckets * sizeof each object.

leelavg commented on services/ux-backend/handlers/featureflags/handler.go:
```diff
@@ -0,0 +1,215 @@
+		return
+	}
+}
+
```
too much chaining of functions increasing the call depth. below is my suggestion:
```
func isNoobaaAvailable() {
// do a get on NoobaaCR, it's not about hardcoding it's about not using the info we already know
// do a get on StorageClass with MatchingField on Provisioner (yes, we need cache + index), again we know the name of Provisioner
}

func isRGWAvailable() {
// ditto as above
}
```
ux-backend is an extension of ocs-op and ocs-op is the one creating or dictating the values for the fields that we are interested in and not some other component, not using them is a missed opportunity.

SanjalKatiyar replied:
https://github.com/red-hat-storage/ocs-operator/pull/3536/files#r2476652372

leelavg commented on services/ux-backend/handlers/bucket/handler.go:
```diff
@@ -0,0 +1,94 @@
+	}
+}
+
+// request format: /auth/bucket/<bucket-name>
```
ack, you can keep it then.

rewantsoni commented on services/ux-backend/handlers/featureflags/handler.go:
```diff
@@ -0,0 +1,214 @@
+	
+	klog.Info("No RGW StorageClasses found")
+	return false, nil
+}
```
I think this comment was not addressed. I still see chaining in the functions that is increasing the call depth. I think you should let the reviewer resolve the comments https://github.com/red-hat-storage/ocs-operator/pull/3536#discussion_r2468005832

rewantsoni replied:
We shouldn't create unnecessary functions if they aren't required. In this case, you are concerned about whether Noobaa/CephRGW is available or not. I don't think we need to split the logic for CR and storageClass into two different functions. I would argue that we aren't benefiting from it; instead, it's making the code less readable and making me jump from one line to another to see what the code is trying to achieve.

leelavg replied:
I don't want to force, you are nack'ng the change. let it be, I don't have anymore supporting arguments.

SanjalKatiyar replied:
Resolving this comment earlier was a mistake on my end.

rewantsoni commented on services/ux-backend/handlers/storages/handler.go:
```diff
@@ -0,0 +1,143 @@
+
+const (
+	deploymentTypeInternal = "internal"
+	deploymentTypeExternal = "external"
```
We already have deploymentType defined in [storagecluster/odfInfo](https://github.com/red-hat-storage/ocs-operator/blob/main/controllers/storagecluster/odfinfoconfig.go#L26-L27), let's move it to the utils package and then use it instead of having new constants

rewantsoni replied:
`deploymentType` Is not something that's going to change even if they do, they have to be changed in both places. They operate independently, but share the same codebase. I think it would be good not to define them again and move them to the utils package

rewantsoni commented on services/ux-backend/handlers/storages/handler.go:
```diff
@@ -0,0 +1,143 @@
+
+	if len(cos.Status.Endpoints.Secure) > 0 {
+		return cos.Status.Endpoints.Secure[0]
+	}
```
```suggestion
	if cos.Status == nil {
		return ""
	}

	if endpoint := cos.Status.Info["secureEndpoint"]; endpoint != "" {
		return endpoint
	}

	if len(cos.Status.Endpoints.Secure) > 0 {
		return cos.Status.Endpoints.Secure[0]
	}
```
Will there be different values at cos.Status.Endpoints.Secure and cos.Status.Info["secureEndpoint"]?

SanjalKatiyar replied:
another example of nit-pickyness :) but I will update it because it's cleaner...

> Will there be different values at cos.Status.Endpoints.Secure and cos.Status.Info["secureEndpoint"]

No, it's just a fallback.

rewantsoni replied:
Can there be a case where we can have one field populated and the other one isn't?

rewantsoni replied:
It would be good to know if there is a case, if there can be a scenario like this, and if yes, then which endpoint should we give a priority to? If it cannot be different, then let's use one of the fields. Why write more code than what is required?

rewantsoni commented on services/ux-backend/handlers/featureflags/handler.go:
```diff
@@ -0,0 +1,214 @@
+func checkNoobaaStorageClassExists(ctx context.Context, client ctrlclient.Client, namespace string) (bool, error) {
+	klog.Info("Checking for NooBaa StorageClass existence")
+	
+	provisioner := fmt.Sprintf("%s.noobaa.io/obc", namespace)
```
provisioner name won't depend on namespace it's always hardcoded https://github.com/red-hat-storage/ocs-operator/blob/main/controllers/util/csi.go#L4-L8

SanjalKatiyar replied:
Yes, I noticed this file.
But, I prefer not to hardcode, unless there is absolutely no other way out.

SanjalKatiyar replied:
Okay, I think there is one case where it can break. What happens in ROSA ?? It is still "openshift-storage", right ?? If so, then probably hardcoding is the only way out.

SanjalKatiyar replied:
Also, [csi.go#L4-L8](https://github.com/red-hat-storage/ocs-operator/blob/main/controllers/util/csi.go#L4-L8) is for Ceph, what about NooBaa, does that use "openshift-storage" prefix in ROSA as well ??

SanjalKatiyar replied:
Another thing, can these provisioner prefix be overridden by users via some spec or env variable or something similar, does that happen in downstream ??

SanjalKatiyar replied:
Or, I keep indexing and if a StorageClass is not found using the index, I fallback to list and filter, just to be sure. But, that is not a clean code, rather a mid-way between 2 approaches which is guaranteed to always work in case of edge cases.

rewantsoni commented on services/ux-backend/handlers/featureflags/handler.go:
```diff
@@ -0,0 +1,214 @@
+func checkRGWStorageClassExists(ctx context.Context, client ctrlclient.Client, namespace string) (bool, error) {
+	klog.Info("Checking for RGW StorageClass existence")
+	
+	provisioner := fmt.Sprintf("%s.ceph.rook.io/bucket", namespace)
```
Provisioner name is hardcoded and doesn't depend on namespace https://github.com/red-hat-storage/ocs-operator/blob/main/controllers/util/csi.go#L4-L8

SanjalKatiyar replied:
https://github.com/red-hat-storage/ocs-operator/pull/3536#discussion_r2476692314

leelavg commented on services/ux-backend/main.go:
```diff
@@ -73,12 +83,68 @@ func main() {
+	}); err != nil {
+		klog.Exitf("failed to add provisioner index for StorageClasses: %v", err)
+	}
 
```
I agree an index for provisioner isn't mandatory and can simply loop over at call site.

SanjalKatiyar replied:
ack, I will remove indexing on StorageClass then. Thanks for confirming.

leelavg replied:
no, it is not a prefix, that's my point in https://github.com/red-hat-storage/ocs-operator/pull/3536#discussion_r2477182388 as mentioned in the linked comment, did you ask either RGW or MCG folks to confirm? I believe they are not changing it, however better confirm once.

leelavg commented on services/ux-backend/main.go:
```diff
@@ -73,12 +83,68 @@ func main() {
+			return []string{ob.Spec.Endpoint.BucketName}
+		}
+		return []string{}
+	}); err != nil {
```
you've mentioned there could be 100 buckets as well, may I known what is the timeout that UI tolerates for the bucket/{name} endpoint?

reason is, by creating a cache we get fast lookup at the cost of increased memory consumption or can UI tolerate the latency of looping through bucket list for each call + num of concurrent users. Since you know the calling pattern better from UI, I want you to decide:

Suppose there are 100 buckets and as each bucket endpoint is unique the indexing overhead is 100 * X for each index value.

Are there any pieces in the bucket name which we can factor out and use a single key as the index, as an example:
b1 - xyz.some.123
b2 - abc.some.345
b3 - asd.some.789

so in all these endpoints `some` is common and we can index on that which doesn't create much overhead.

SanjalKatiyar replied:
Default timeout should be 60s.

UI should tolerate IMO. Still, as per my latest/updated understanding cache is lazily loaded (plz correct me if I am wrong), and we will only need to call this endpoint in those clusters where users have non-admin use case. For users with admin only use cases, those clusters will never call and hence populate the cache with ObjectBuckets.

IIUC, your main concern here is overhead of the index itself ?? Sadly, I don't see any other way for indexing, bucket name is the only information that UI has and it can send in the request to uniquely identify the bucket. But, if u prefer then just like StorageClasses case, we can remove indexing here as well, let me know ??

leelavg replied:
> understanding cache is lazily loaded
- since we don't have watches when a request is made to list bucket only then cache is initially synced with api server for the first time, pls do note this'll incure additional time during first call based on number of buckets

last question:
does the UI use this data in a list mode or individual item mode, like, when a page is displayed with all the buckets does that have a icon/info representing the creation method or it gets displayed only when users drills into a specific bucket.

let's go w/ current implementation and later on (after FF) we can do some simple test to see the memory growth?

SanjalKatiyar commented on services/ux-backend/main.go:
```diff
@@ -73,12 +83,68 @@ func main() {
+	}
 
-	cl, err := util.NewK8sClient(scheme)
+	if err := cache.IndexField(ctx, &nbv1alpha1.ObjectBucket{}, "bucketName", func(obj client.Object) []string {
```
make sense... will update...

SanjalKatiyar commented on services/ux-backend/handlers/bucket/handler.go:
```diff
@@ -0,0 +1,94 @@
+	klog.Infof("Checking if bucket '%s' was created via OBC", bucketName)
+	
+	objectBucketList := &nbv1alpha1.ObjectBucketList{}
+	if err := client.List(ctx, objectBucketList); err != nil {
```
If by "bucketname" name u mean ObjectBucket CR name, then yes. Underlying actual s3 bucket (in spec) has different name than the CR name.

leelavg commented on services/ux-backend/handlers/bucket/handler.go:
```diff
@@ -0,0 +1,86 @@
+
+	"github.com/red-hat-storage/ocs-operator/v4/services/ux-backend/handlers"
+
+	nbv1alpha1 "github.com/noobaa/noobaa-operator/v5/pkg/apis/noobaa/v1alpha1"
```
```suggestion
	nbv1a1 "github.com/noobaa/noobaa-operator/v5/pkg/apis/noobaa/v1alpha1"
```
nit, change if you are fine & refreshing the PR.

SanjalKatiyar commented on config/rbac/oauth_proxy_role.yaml:
```diff
@@ -10,10 +10,34 @@ rules:
+  - storageclusters
+  verbs:
+  - get
+  - list
```
Yup, I have yet to update the PR. Thanks for pointing it out, will update and test soon.

leelavg commented on services/ux-backend/handlers/bucket/handler.go:
```diff
@@ -0,0 +1,86 @@
+		return creationMethodOBC, nil
+	}
+	
+	klog.Infof("No ObjectBucket found with bucket name '%s', assuming S3 creation", bucketName)
```
the comment isn't helpful. it should mention why we consider the absence of a obc resource as created by s3, it's due to UI calling this api against an existing bucket.

leelavg reviewed (commented):
only 2 points remain for me:
1. you only have one noobaa cr with a specific name and list wouldn't capture that intent, sometimes it's erroneous if indeed you have 1 noobaa cr but with a different name and mcg woudl refuse to reconcile it
2. matching the provisioner name by prefix, it doesn't matter what the environment is, it's always `openshift-storage.<our-driver>` pls note this is not a prefix, I just wrote it as there are many drivers. if there is `<something>.<our-driver>` you'd consider that such a storageclass exists which is wrong. (any comment on this statement @Madhu-1? thanks).

SanjalKatiyar commented on rbac/ux_backend_role.yaml:
I doubt so. I will test and add if I find any missing RBACs.

`storageclusters` already has cluster-wide list/watch privilege, this file here only adds extra `update` in `openshift-storage` namespace (no need for cluster-wide `update` yet).

`cephblockpools` are also only allowed `create` for now, list/watch not needed yet.

`secrets` are not fetched via endpoint's cached client afaik, used only by the service account.

SanjalKatiyar commented:
/hold
(need to test once)

SanjalKatiyar commented:
/hold cancel

SanjalKatiyar commented:
@leelavg @rewantsoni addressed all comments and did the final testing (updated the PR description with relevant testing details).
PTAL and tag (if it looks good). Thanks !!
